File: lib/src/agent/tool-arg-inference.ts
export function inferToolArgs(toolName: string, userMessage: string): Record<string, unknown> | null {
  const msg = userMessage.toLowerCase();

  switch (toolName) {
    case 'calculator': {
      const mathMatch = msg.match(/([\d.]+)\s*([+\-×÷*/xX]\s*[\d.]+)+/);
      if (mathMatch) {
        return { expression: mathMatch[0].replace(/[xX]/g, '*').replace(/[÷]/g, '/') };
      }
      const numMatch = msg.match(/([\d.]+)\s*(더하기|빼기|곱하기|나누기|plus|minus|times|divided)\s*([\d.]+)/i);
      if (numMatch) {
        const op = numMatch[2]
          .replace('더하기', '+').replace('plus', '+')
          .replace('빼기', '-').replace('minus', '-')
          .replace('곱하기', '*').replace('times', '*')
          .replace('나누기', '/').replace('divided', '/');
        return { expression: `${numMatch[1]}${op}${numMatch[3]}` };
      }
      return null;
    }

    case 'notepad': {
      const memoMatch = userMessage.match(/(?:메모|기록|저장|memo|note|save)[:\s]*(.+)/i);
      if (memoMatch) return { action: 'save', key: 'memo', value: memoMatch[1].trim() };
      return null;
    }

    case 'timer': {
      const minMatch = msg.match(/(\d+)\s*분/);
      if (minMatch) return { action: 'set', seconds: parseInt(minMatch[1]) * 60 };
      const secMatch = msg.match(/(\d+)\s*초/);
      if (secMatch) return { action: 'set', seconds: parseInt(secMatch[1]) };
      return null;
    }

    default:
      return null;
  }
}

File: lib/domain/agent/tool_arg_inference.dart
Map<String, dynamic>? inferToolArgs(String toolName, String userMessage) {
  final msg = userMessage.toLowerCase();
  switch (toolName) {
    case 'calculator':
      final expr = extractMathExpr(msg);
      if (expr != null) return {'expression': expr};
      return null;
    case 'notepad':
      final writeMatch = RegExp(
        r'(?:write|save|note|store|record)\s+(.+)',
        caseSensitive: false,
      ).firstMatch(userMessage);
      if (writeMatch != null) {
        return {
          'action': 'write',
          'key': 'note_${DateTime.now().millisecondsSinceEpoch}',
          'content': writeMatch.group(1)!.trim(),
        };
      }
      return {'action': 'list'};
    case 'timer':
      final durationMatch = RegExp(
        r'(\d+)\s*(?:second|sec|minute|min)',
        caseSensitive: false,
      ).firstMatch(msg);
      if (durationMatch != null) {
        final value = int.tryParse(durationMatch.group(1)!) ?? 0;
        final unit = msg.contains('min') ? value * 60 : value;
        return {'action': 'set', 'seconds': unit};
      }
      return null;
    default:
      return null;
  }
}

String? extractMathExpr(String msg) {
  final ops = {
    'plus': '+',
    'added to': '+',
    'and': '+',
    'minus': '-',
    'less': '-',
    'subtract': '-',
    'times': '*',
    'multiplied by': '*',
    'x': '*',
    'divided by': '/',
    'over': '/',
  };
  var expr = msg;
  expr = expr.replaceAll(RegExp(r'calculate\s*', caseSensitive: false), '');
  expr = expr.replaceAll(RegExp(r'what\s+is\s*', caseSensitive: false), '');
  expr = expr.replaceAll(RegExp(r'compute\s*', caseSensitive: false), '');
  for (final entry in ops.entries) {
    expr = expr.replaceAll(entry.key, entry.value);
  }
  expr = expr.replaceAll(RegExp(r'[^\d+\-*/.()% ]'), '').trim();
  if (expr.isEmpty || !RegExp(r'\d').hasMatch(expr)) {
    return null;
  }
  if (!RegExp(r'[+\-*/]').hasMatch(expr)) return null;
  return expr;
}
